[benchmark] Naming Convention

palimondo opened the pull request:
After short discussion with @lorentey [about benchmark names](https://github.com/apple/swift/pull/20067#issuecomment-434663394) and my work on re-enabling `ExistentialPerformance` benchmarks and looking at how to shorten the names of `ObjectiveCBridge*`benchmarks in the future, I'm proposing  a new benchmark naming convention for better readability and improved naming system that accounts for performance coverage growth going forward.

The PR was updates following a lengthy conversation during review, [see comments](https://github.com/apple/swift/pull/20334#discussion_r230956365), to clarify use of abbreviations.

----

To create more cohesive and well structured system, names of newly added benchmarks should meet the following set of requirements:

* **Letters, numbers and dots.** Start with short unique name in `UpperCamelCase`. For a family of benchmarks, separate the name components with periods.
* **Use groups and variants** to structure the benchmark family by degrees of freedom (e.g. different types implementing a protocol, value vs. reference types etc.). Use numbered suffixes to denote differently sized variants.
* **Groups and types are `UpperCase`, methods are `lowerCase`.**
* **Keep it short.** 40 characters at most. Abbreviate if necessary.

Technically, the benchmark's name must match the following regular expression:
`[A-Z][a-zA-Z0-9\-.!?]+`

More details with examples are in the [Naming.md](https://github.com/apple/swift/blob/419ec8b238364439eda5527bb6e1bcfdf92b3451/benchmark/Naming.md) document.

palimondo committed:
[benchmark] Naming Convention

New benchmark naming convention for better readability and improved naming system that accounts for performance coverage growth going forward.

palimondo commented:
Please review: @eeckstein, @gottesmm, @atrick, @airspeedswift, @lorentey, @dabrahams

jrose-apple commented on benchmark/Naming.md:
```diff
@@ -0,0 +1,148 @@
+
+In a pinch, even C-style naming with two letter prefixes `OC` (for Objective-C)
+or abbreviations like  `Str` and `Arr` are OK, if it helps to fit a system with
+descriptive names into 40 characters:
```
I don't work with the benchmarks much, but this doesn't match either LLVM or Swift naming conventions. I'd rather support having the benchmark runner insert ellipses into long names than telling people to start dropping vowels or syllables.

atrick replied:
This about coding convention, it's about readability of the report where we have a very long list of repetitive names. Abbreviations work well here.

jrose-apple replied:
But we don't usually read the long list of names; only the ones that change come up.

lorentey replied:
I agree with Jordan. I think it's okay to use abbreviations that are relatively widespread in the codebase anyway (e.g., ObjC for Objective-C), but inventing new ones hurts readability a lot:

Current: 

> ObjectiveCBridgeFromNSArrayAnyObjectForced

Better (although still not great):

> ObjectiveC.Bridge.From.NSArray.AnyObject.Forced 

Just as good: 

> ObjC.Bridge.From.NSArray.AnyObject.Forced

Worse:

> OC.Bridge.From.NSArr.AO.Forced

I hope we don't currently have anything as egregious as that last one.

atrick replied:
Fine with me, as long as parts of name can be easily elided in text-based tabular output.

palimondo replied:
The `ObjectiveCBridge*` family of benchmarks is a good illustration of how to correctly apply the new principles. It has made me reconsider the legal set of characters to include `?` and `!`. The abbreviations I had in mind in that case @lorentey mentions would just mean replacing `ObjectiveCBridge` as a family prefix with `OCB`:
````
⛔️ ObjectiveCBridgeFromNSArrayAnyObjectForced
✅ OCB.NSArray.AnyObject.as!.Array.NSString
````

The details in the first bullet include it other members from that family in the example:
````
✅ OCB.NSArray.AnyObject.as?.Array.NSString
✅ OCB.NSArray.AnyObject.as!.Array.String
````
The point of **Keep it short** is to not be afraid to abbreviate where necessary, in order to fit the family into reasonable size, but to remain consistent. It's also illustrated in the details of the *Groups and variants* bullet.
````
✅ Seq.dropFirst.Array
✅ Seq.dropLast.Range.lazy
✅ Seq.dropWhile.UnfoldSeq
✅ Seq.prefix.AnySeq.RangeIter.lazy
✅ Seq.prefixWhile.AnyCol.Array
✅ Seq.suffix.AnySeq.UnfoldSeq.lazy
````
The local abbreviation rules in that benchmark family look like this:

| Type | Abbreviation |
|---|---|
| `AnyCollection` | `AnyCol` |
| `AnySequence` | `AnySeq` |
| `UnfoldSequence` | `UnfoldSeq` |
| `IndexingIterator<(Range<Int>)>` | `RangeIter` |

Should I rephrase the last point? Add more examples? I thought the document-as-is already contains it all, but apparently it isn't clear enough...

palimondo replied:
The ideas about groups, name components and consistent abbreviations are lessons learned from introducing [extended coverage for sequence slicing operations](https://github.com/apple/swift/pull/8641). I tried my best with sequence names at the time, but when you pull all the results for comparison, [like I did here](https://forums.swift.org/t/troubling-sequence/13130) (look for table in under section **Cost of Abstractions, Revisited**), the newly proposed system can be used to slide and dice such a family much easier. When pulled together with new naming convention, the `Seq` family has 3 degrees of freedom: slicing **operation**, underlying sequence **type** and the eager/**lazy** dichotomy.

lorentey replied:
The problem with using ad-hoc abbreviations like `OCB` is that they don't mean anything at all.

* "ObjectiveCBridgeFromNSArrayAnyObjectForced" is quite clumsy and not very readable, but if it pops up in a list of benchmark regressions I can look at the name and have a rough picture of what it measures.

* "OCB.NSArray.AnyObject.as!.Array.NSString" reads like an opaque mix of random characters and type names to me. I can see it has something to do with arrays and strings, but that's about it. I have to look up the source to see what it does.

The sequence slicing names are interesting; the abbreviations and camel casing do work pretty well together there. It hlps that the abbrevs are rltvly understable. But writing out the actual types in informal Swift syntax would still be a huge improvement.

I'm increasingly convinced these rules just get in the way of clarity. The goal is to make it clear what the benchmark measures -- so why not just write that down in whatever notation feels best?

```
ObjectiveCBridgeFromNSArrayAnyObjectForced
      → Bridging: NSArray<AnyObject> as! Array<NSString>

ObjectiveCBridgeStubFromArrayOfNSString2
      → Bridging: Stub from NSArray<NSString> (v2)

DropLastAnySeqCntRange
      → DropLast: AnySequence(CountableRange<Int>)
   or → AnySequence(CountableRange<Int>).dropLast()
    
DropFirstAnySeqCRangeIterLazy
      → DropFirst: AnySequence(CountableRange<Int>.Iterator).lazy
   or → AnySequence(CountableRange<Int>.Iterator).lazy.dropFirst()

RomanNumbers
      → Algorithm: Roman numeral conversion to/from Int
```

palimondo replied:
I'm not sure we are trying to solve the same problem… The size of the benchmark names, when it comes to big families like `Existential` or `ObjectiveCBridge` are simply too long, and are making the reports unusable because of the horizontal scrolling. Names are used to invoke the individual benchmarks from the benchmark driver. I don't see how what you proposed(?) above addresses any of these issues.

lorentey replied:
We definitely aren't! :) 

*My* pet problem is that benchmarks aren't labeled clearly, and I don't want them to get even worse. This is why I pushed back on enforcing camel casing, and this is why I dislike a-h abs & SATN. After all, TISNAEWTC!

On the other hand, having to scroll around on a webpage, or having to sometimes quote things on the command line honestly seem relatively minor issues to me. It is possible to design around those, while bad names remain bad no matter how well we display them.

palimondo replied:
I'm working on integrating the [benchmark validation](https://github.com/apple/swift/blob/a7f832fb57b459dc07c577db551bf88b39fa0629/benchmark/scripts/Benchmark_Driver#L264-L489) into the benchmark [command for CI](https://github.com/apple/swift/pull/20074). It checks that that **newly added** benchmarks conform to the best practices. For deviations it offers suggestions how to correct the issues.

Historically, one of the issues hampering usability of Swift Benchmark Suite were overlong test names, that were causing:
* misaligned columns in console output when running benchmarks with `build_script` and
* hiding the most important result columns in benchmark report tables on GitHub.

I have been trying to fix this problem for [more than year and a half](https://github.com/apple/swift/pull/9999#issuecomment-305296057). Back than, @atrick agreed that 40 characters would be a [reasonable limit:](https://github.com/apple/swift/pull/9999#issuecomment-305298530)
> The long benchmark names ruin my ability to table-ify the output, independent of Markdown. If benchmarks need verbose descriptions, we should add those. 40 chars is already ridiculous for a benchmark name.

Which is why  `BenchmarkDoctor`'s `check` has now a rule for [name length](https://github.com/apple/swift/blob/a7f832fb57b459dc07c577db551bf88b39fa0629/benchmark/scripts/Benchmark_Driver#L324-L331). In order to maintain consistency among benchmarks names, I've also added rule for [naming convention](https://github.com/apple/swift/blob/a7f832fb57b459dc07c577db551bf88b39fa0629/benchmark/scripts/Benchmark_Driver#L311-L321), pointing to the `UpperCamelCase` in Swift Naming Guidelines.

@lorentey, you fairly [pushed against that](https://github.com/apple/swift/pull/20067#issuecomment-434663394) and [said (emphasis mine)](https://github.com/apple/swift/pull/20067#issuecomment-435482953):
> To me, **any separator** would be **much preferable** to `DictionaryOfAnyHashableStringsInsert`. `DictionaryOfAnyHashableStrings` is already quite bad, but appending Insert to it makes it unreadable.
> 
> Benchmark names are very similar to test names -- some of them are **hierarchical** we and **should have a rule** that accommodates that. **The dot would work very well!**
> 
> Names like `SetSymmetricDifferenceInt50` didn't bother me as much because their components are more straightforward, but `Set<Int>.symmetricDifference() 50%` would still read a lot better in any table. However, `Set.Int.symmetricDifference.50` is a **reasonable compromise** between the two, and it's still quite usable.

Incorporating your feedback I've proposed this naming convention, so that the `BenchmarkDoctor` can point to a concrete set of rules.

I hear that your main issue with benchmarks is clarity on what each of them does. Which is also fair, but I don't think it's realistic expectation to fully understand a benchmark just from its name. It's often the case that even looking at the source code leaves you without full comprehension of the original author's intent. Which is why @atrick's been pushing me to be [more diligent about documentation](https://github.com/apple/swift/pull/20067#issuecomment-435482953):

> For future benchmark maintenance, it's important to understand the justification and intention behind the benchmarks. Please add comments to test cases explaining what is being tested and why we care about performance of this particular test vs. more conventional ways to write it. I know nobody has really done this in the past, but I think it's important, especially when adding such a large number of strangely written tests.

Back to the proposed naming convention: the objections raised seem quite knee-jerk to me. It would take exactly one look at ObjectiveCBridging.swift to understand what `OCB` stands for. Repeating `ObjectiveCBridging` in the name of each test just clutters everything.

Allowing `?` and `!` in names is shortest notation describing how tests vary across two degrees of freedom — types being cast and optional/forced downcast. Avoiding the `Forced` suffix, allows adding **more information** about what and how is being tested to the name. Finishing the hypothetical renaming of tests from ObjectiveCBridging.swift according to proposed naming convention, I've realized that inclusion of `AnyObject` in the names was not adding information and caused name length overflow. Abbreviating `NSDictionary` to `NSDict` and `NSNumber` to `NSNum` made `OCB.NSDict.as!.Dictionary.NSString.NSNum` fit exactly the 40 characters limit. Here's the final result:

| After | Before |
|---|---|
| `OCB.NSString.as?.String` | `ObjectiveCBridgeFromNSString` |
| `OCB.NSString.as!.String` | `ObjectiveCBridgeFromNSStringForced` |
| `OCB.String.as.NSString` | `ObjectiveCBridgeToNSString` |
| `OCB.NSArray.as?.Array.NSString` | `ObjectiveCBridgeFromNSArrayAnyObject` |
| `OCB.NSArray.as!.Array.NSString` | `ObjectiveCBridgeFromNSArrayAnyObjectForced` |
| `OCB.Array.String.as.NSArray` | `ObjectiveCBridgeToNSArray` |
| `OCB.NSArray.as?.Array.String` | `ObjectiveCBridgeFromNSArrayAnyObjectToString` |
| `OCB.NSArray.as!.Array.String` | `ObjectiveCBridgeFromNSArrayAnyObjectToStringForced` |
| `OCB.NSDict.as?.Dictionary.NSString.NSNum` | `ObjectiveCBridgeFromNSDictionaryAnyObject` |
| `OCB.NSDict.as!.Dictionary.NSString.NSNum` | `ObjectiveCBridgeFromNSDictionaryAnyObjectForced` |
| `OCB.Dictionary.String.Int.as.NSDict` | `ObjectiveCBridgeToNSDictionary` |
| `OCB.NSDict.as?.Dictionary.String.Int` | `ObjectiveCBridgeFromNSDictionaryAnyObjectToString` |
| `OCB.NSDict.as!.Dictionary.String.Int` | `ObjectiveCBridgeFromNSDictionaryAnyObjectToStringForced` |
| `OCB.NSSet.as?.Set.NSString` | `ObjectiveCBridgeFromNSSetAnyObject` |
| `OCB.NSSet.as!.Set.NSString` | `ObjectiveCBridgeFromNSSetAnyObjectForced` |
| `OCB.Set.String.as.NSSet` | `ObjectiveCBridgeToNSSet` |
| `OCB.NSSet.as?.Set.String` | `ObjectiveCBridgeFromNSSetAnyObjectToString` |
| `OCB.NSSet.as!.Set.String` | `ObjectiveCBridgeFromNSSetAnyObjectToStringForced` |

Overall, I don't find the suggestion to allow free form names with spaces, brackets and braces remotely practical. Even allowing for `.` and `-` in the names [requires some modifications to the scripts that run benchmarks](https://github.com/apple/swift/pull/20667)… I admit that allowing `!` in the names would cross a line and require escaping (`\!`) or quoting the name when running such test individually in the command line. But it's still more practical and much more constrained design than freeform strings!

The whole workaround to overlong names of not showing all benchmark results in the CI reports on GitHub is IMHO a major regression in the transparency of Swift project. It's nice that you have Apple-internal performance tracking available, but since the the introduction of `run_smoke_bench`, there is no way for outside contributors to monitor performance of the Swift project without running the benchmarks locally. Previously I could search GitHub for the benchmark reports and get a sense of what's going on. I find smoke bench fine for what it does, but I think the ability to get full performance reports from CI should be re-established. Fixing the naming is therefore still important. Enforcing best practices for newly added tests is a first step.

Let's not throw out the baby with the bathwater here… What modifications to the proposed naming convention would make all of you agree that adding this to the project is beneficial?

atrick replied:
> I have been trying to fix this problem for more than year and a half. Back than, @atrick agreed that 40 characters would be a reasonable limit:

I also said that we can make do as long as there's some effective way to elide parts of the name. Once you have hierarchically organized tests and name separators, that limit would only apply to individual components of the name. I regularly parse and reformat long lists of tests in tabular text format. As long as there's a way to format the text output such that it's actually readable and can be automatically converted to a table, that works for me. 

e.g. output like this work work org mode's table-convert-region.
```
NAME                VAL, VAL, VAL
Bench1Group.
Bench1Name     #,  #,  #
Bench2Group.
Bench2Name     #, #, #
```

palimondo replied:
I'll rephrase the explanation to stress the last resort nature of using abbreviations:

> *In a pinch*, even C-style naming with two letter prefixes `OCB` (for
Objective-C Bridging) or consistent abbreviations in the benchmark family like
`Str` and `Arr` are OK, but *only if* it helps to fit a system with descriptive
names into 40 characters:
> 
> ````
> ✅ CharCount.ContArr.Str.kr
> ✅ OCB.NSDict.as!.Dictionary.NSString.NSNum
> ✅ Seq.prefixWhile.AnySeq.UnfoldSeq.lazy
> ````

palimondo replied:
Not restricting the length of a name is generally problematic, as it complicates effective design of other tools. Here’s the visualization of the whole benchmark suite runtimes I have been using to asses the health of the SBS: 

![benchmarks.html](https://user-images.githubusercontent.com/1046899/48694772-95e65480-ebdd-11e8-8f49-deb59edc810c.jpg)

In contrast to my [chart.html visualizations](https://palimondo.github.io/robust-microbench/chart.html?b=Calculator&v=iters), which work well on any mobile device thanks to responsive design, the overlong names are making benchmarks.html hard to use on smaller screens.

palimondo committed:
[benchmark] Simpler naming convention regex

palimondo committed:
[benchmark] Naming: emphasize ABBR are last resort

jrose-apple commented:
I'm staying out of this at this point because I don't use the benchmarks nearly as much as the people who more frequently work in the stdlib or on optimization passes, but I still think [clarity is more important than brevity](https://swift.org/documentation/api-design-guidelines/#fundamentals) and that inventing non-standard acronyms like "OCB" is just going to confuse people.

xwu commented:
I have to agree that the proposed short names here areessentially unreadable to me compared to the existing names. They may make sense to someone who reads these regularly but since many more people will be reading only the few benchmark names that are affected by their changes these would be equivalent to showing the first 40 characters of the SHA1 hash of a benchmark.

palimondo commented:
@jrose-apple, @dabrahams @xwu, I'm not arguing for the return of C-style naming! Don't let the allergic reaction to sins past cloud your judgment, please...

> proposed short names here are essentially unreadable to me compared to the existing names

Assuming you're comparing the `ObjectiveCBridge` to `OCB` examples from [the discussion](https://github.com/apple/swift/pull/20334#discussion_r234462042), could you examine what's your impression actually based on? 

Ignoring `OCB` for the moment, the explanatory power in the rest of the name is actually improved. It includes more of the relevant information in a more efficient form:
 
Before | After
-- | --
ObjectiveCBridgeFromNSDictionaryAnyObject | OCB.NSDict.as?.Dictionary.NSString.NSNum 
ObjectiveCBridgeFromNSDictionaryAnyObjectForced | OCB.NSDict.as!.Dictionary.NSString.NSNum 
ObjectiveCBridgeFromNSDictionaryAnyObjectToString | OCB.NSDict.as?.Dictionary.String.Int
ObjectiveCBridgeFromNSDictionaryAnyObjectToStringForced | OCB.NSDict.as!.Dictionary.String.Int

`OCB` — a **family name** in the hypothetical rename — serves just as a binder that holds all these tests together as a logical group. 

> I still think [clarity is more important than brevity](https://swift.org/documentation/api-design-guidelines/#fundamentals) and that inventing non-standard acronyms like "OCB" is just going to confuse people.

I've argued that case before:

> I hear that your main issue with benchmarks is clarity on what each of them does. Which is also fair, but I don't think it's realistic expectation to fully understand a benchmark just from its name.

But I think you are all highlighting a real issue here: making the benchmark reports more actionable for someone without intimate knowledge of the Swift Benchmark Suite. Or as Xiaodi put it:

> They may make sense to someone who reads these regularly but since many more people will be reading only the few benchmark names that are affected by their changes these would be equivalent to showing the first 40 characters of the SHA1 hash of a benchmark.

Which, to me, actually hints at a better solution… Benchmark **names alone**, even though they contain a pinch of semantic hints about their purpose, **are not enough** to take reasonable action when they show up in a report on GitHub. Lifting length restrictions on benchmark names doesn't solve that problem. Benchmark names — like the commit hashes — are just unique identifiers. You need to **consult** benchmark's actual **source code anyway**.  That's what needs to be made easier!

Hyperlinking is what makes SHA1 hashes usable on GitHub. I can extend `BenchmarkInfo`s with file names and line numbers, print these from `Benchmark_O` with `--list=src` and use that to construct links to actual source code on GitHub for each benchmark name in the Markdown reports.

Limiting benchmark names to fixed length greatly eases creation of existing and all future tooling for visualizing our performance. I have several ideas, especially for relative comparison of variants and groups inside a single family.

atrick commented:
@palimondo Thanks for doing this.

I've never understood the argument for absurdly long camel case names. When you have many of these similar names scattered throughout the data set, it becomes impossible for me to visually distinguish them from each other, and I often get them mixed up. Brevity increases clarity. Using namespaces and separators also helps a lot.

You're right, the purpose of these names is simply to have unique benchmark IDs. The reason we don't use numbers is that it also helps to have something memorable. They should to easy to pick out in a large data set. They *need* to work with all manner of data formatting tools. I often *manually* elide large names just so I can more easily work with the data table.

People seem to be arguing that the name should indicate the purpose of the benchmark, which is just ridiulous. You know what would actually help to quickly communicate the purpose of a benchmark: benchmark descriptions and comments in the benchmark! No one says we can't output the short description in whatever web view people are arguing about.

nit: "OCB" is a rather bizarre choice given that "ObjC" is an accepted abbreviation. Why does "Bridge" need to be in the name at all? It can be in the description!

palimondo commented:
> nit: "OCB" is a rather bizarre choice given that "ObjC" is an accepted abbreviation. Why does "Bridge" need to be in the name at all? It can be in the description!

There's nothing religious about the choices I made… Considering the longest name `OCB.NSDict.as?.Dictionary.NSString.NSNum`, **`OCB`** was just an acronym that made it all fit under 40 characters while still spelling `Dictionary` in full. With `ObjC` it is one character too long, forcing abbreviation — `Dict` — but then we can be more descriptive with the family name:  **`Bridging`**`.NSDict.as?.Dict.NSString.NSNum`. I think that actually bolsters the point about consistent intra-family abbreviation. I'll update Naming.md to match.

Are there any other fundamental objections to the proposed naming convention?

palimondo committed:
[benchmark] Replace OCB with Bridging

Removed the controversial mention of C-style name prefixes.

palimondo commented:
Also, any feedback on the idea to provide links to source code from benchmark reports?
<details>
<summary>Implementation details</summary>

While the idea with extending `BenchmarkInfo` with `#file` and `#line` would be workable, it can link only to the declarations — not the `runFunction` implementation. It's always necessary to search for the runFunction within the file. So I guess equally good solution without modifying `BenchmarkInfo` is to generate GitHub search links from the name like this:

[ExistentialTestArrayTwoMethodCalls_ClassValueBuffer1](../search?q=name+"ExistentialTestArrayTwoMethodCalls_ClassValueBuffer1"+runFunction)

</details>

lorentey commented:
Allowing periods as separators is definitely an improvement over the status quo. Adding free-form descriptions is a good idea, too.

gottesmm commented:
@atrick I am fine with the namespace separator. I think having something like that makes sense and provides a good way to slice/dice results. That being said I do object to this PR in its current form since I still do not understand why the 40 character limit is the only way to solve the problem that Pavol is trying to solve. Specifically:

a. The argument around HTML tables is a non-problem since we can always put the test names on the right hand side and that problem is gone. Then one will have:

| DATA_COL | DATA_COL | DATA_COL | TEST_NAME |

b. Visualizations. AFAIKT this is the other argument for the 40 character limit. I do not buy that the 40 character limit is the only solution for this. Why couldn't we instead limit the size of the individual namespaces (perhaps to 10-15 characters) but allow for the user to have as many namespaces as they want. If one looks at the visualization that @palimondo posted above the important thing is the shortness of the namespace names... not the total length. If that suffices then having the limit on the individual namespaces IMO would be a more flexible solution and would stop the wastage of engineering time by people trying to compress names to fit within 40 characters.

Before the naming convention lands, I would at least like my concerns here answered. Why is it that the 40 character limit is the only solution here?

That being said I would be ok with the test renames here landing, as long as the naming convention document and enforcement of the naming convention is removed from the PR.

gottesmm commented:
As an aside, I also think in a follow on commit it would be useful to:

1. Provide a one line description appropriate to dump in a list containing all tests.
2. A long length man page like description of a benchmark, what it is supposed to do, its design constraints, etc.
3. Allow for the user to specify namespaces programmatically via using a hierarchical dictionary with string keys. The way this would work is that string keys with sub-dictionaries, array of BenchmarkInfo, or single BenchmarkInfo are namespaces. By having multiple levels of sub-dictionaries one gets a natural tree effect providing sub-namespaces. We could use this to be able to programmatically work with namespaces to both verify at runtime that namespaces are the appropriate length and fit all conventions/etc and output them in various ways for tooling to use.

gottesmm commented:
Andy and I spoke about this offline, I withdraw my objections.

atrick commented:
I don't think there are any objections to the content of this PR itself. Note that the 40 character name limit was already there. It's not being introduced here.

We haven't fully resolved how to print benchmark descriptions--I can see a couple a viable options and opinions on it are split--but that doesn't need to be discussed in this PR.

palimondo committed:
[benchmark] Use for ISO 639-3 language variants 

Matched the part of the ISO standard used for language code to match one used on https://www.unicode.org/udhr/translations.html

(I plan to use the Universal Declaration of Human Rights as corpus for string testing.)

palimondo commented:
@swift-ci please test

palimondo commented:
@swift-ci please test

atrick reviewed (approved):
LGTM

palimondo commented:
Thank you @atrick and all who participated in the review to make this proposal better! 🤝

xwu commented on benchmark/Naming.md:
```diff
@@ -130,8 +130,8 @@ legacy-style).
 that measure `String`'s Unicode performance for various languages, use
-[two letter codes](https://en.wikipedia.org/wiki/ISO_639-1) instead of spelling
-out the whole language names.
+the [three letter codes](https://en.wikipedia.org/wiki/ISO_639-3) instead of
```
These three-letter codes are pretty esoteric as compared to two-letter codes. I’ve pretty much never seen them used. What’s the thinking behind this change?

palimondo replied:
From commit description:
> Use <del>for</del> ISO 639-3 <ins>for</ins> language variants 
> Matched the part of the ISO standard used for language code to match the one used on 
> https://www.unicode.org/udhr/translations.html
> 
> (I plan to use the Universal Declaration of Human Rights as corpus for string testing.)

xwu replied:
The two-letter codes don’t correspond to countries for either Chinese or English, two of the most common languages, so I doubt many people suffer from such a deception. I’m concerned that the three-letter codes, besides being much less common, have been critiqued as sometimes pejorative to the languages they’ve been assigned and that they are overseen by an opaque process internal to a missionary organization.

palimondo replied:
🤦‍♂️ I'll be posting in the forums about string benchmarks soon, so you can share your concerns there.

palimondo replied:
@xwu FYI: https://forums.swift.org/t/more-representative-string-benchmarks/18793
